Don't warn that a valid version spec is invalid

behrmann opened the pull request:
Currently `systemd-analyze` from the main branch (not in 253) warns
```console
$ systemd-analyze compare-versions 123~post1 123.a-1
Version string 1 is not valid, comparing anyway: 123~post1
123~post1 < 123.a-1
```
which is harmless, but the string should be valid, since it's taken directly from a comment of `strverscmp_improved`.

Looking at `version_is_valid` there's a few characters that are not allowed by the version spec, but beside tilde it's also missing the caret character, so this adds a second function to check for validity against the version spec and uses this in compare-versions.

<!-- devel-freezer = {"comment-id":"1633881525","freezing-tag":"v254-rc1"} -->

bluca reviewed (approved):
Are there other places where this should also be used?

keszybz commented:
I think the commits should be combined. They don't make sense separately, and cannot even be really reviewed.

The spec is very clear:
> The version string is a sequence of zero or more characters.

But this patch rejects `''`. I think it should be allowed, esp. that we had explicit support for that (c20b2f2421011a2c29fa04cc1955ed67b87daeb2).

---

Hmm, ffe5aba60c970ae0a01285215a200b042a6203ff added this warning. Maybe we should change this to something more meaningful, like "Version string 1|2 contains disallowed characters, they will be treated as separators." ?

behrmann commented:
Pushed a new version addressing the feedback. PTAL.

> Are there other places where this should also be used?

I grepped around and I don't think so, but I'm note entirely sure. `version_is_valid` seems to only be used in sysupdate besides analyze (where it's not used any longer).

poettering commented:
hmm? i don't follow here. what's a versionspec? why have two functions for this? sysupdate compares versions with strverscmp_improved() too, so why distinguish anything here? why allow an empty version, that's broken?

behrmann commented:
> hmm? i don't follow here. what's a versionspec? why have two functions for this? 

Sorry for being slightly incomprehensible there. I've been writing "version string following the version specificaton" slightly too often in the past few days and probably shortened it too much.

> sysupdate compares versions with strverscmp_improved() too, so why distinguish anything here? 

On first glance I didn't see right away what sysupdate's needs for `version_is_valid` were, but the allowed character set in there has some things in there that are not allowed by the version spec and is missing some that are (ASCII letters + digits and `.,_-+` vs ASCII letters + digits and  `-.~^`) - I guess for the matching support?

If this can be unified in some way, I'm fine with that, too.

> why allow an empty version, that's broken?

The version spec allows empty strings as versions, as @keszybz pointed out, so generally comparing two versions, this probably shouldn't warn on empty versions.

behrmann commented:
I have no real preference there and if there's agreement I'll happily change this. The first version of this PR still warned on empty version, and this is more about not warning if there are `~` or `^` in the version and to a lesser degree about warning about characters that the spec does not allow.

keszybz commented:
I restarted the rpm-build tests. Centos SI (ArchLinux), jammy-amd64 and 
ci/semaphoreci/pr: Debian autopkgtest (LXC) failed, but seem unrelated.

keszybz commented:
This should be merged for v254, because it fixes a regression in `systemd-analyze compare-versions`. The fix is minimal, and clearly only affects that code.

> This sounds like a bug in the spec. An empty version should not be a valid version

This is a matter of consistency. The versionspec (after my updates) is iterative: strings are compared from the left to right, chunk by chunk, until either they compare different or both are finished. If we get to the end of one of the strings, we need to compare the empty suffix with the remaining part from the other string. If an empty string is allowed, we don't need a special exception for this case.

In practice, for example this means that if we have `foo` and `foo.version`, we can still compare the two, without creating an exception. First we compare `foo`, which is the same in both strings, and then we compare the empty string with the suffix, and get some result according to the usual rules.

Dunno, I think we revisit this question, but I think the spec is good as is. It's totally OK to require non-empty version numbers in some specific cases, but this should be part of the specification of those specific cases, and I think the versionspec itself should remain agnostic.
